Update box model to border box

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Why we need update
box-sizing supported by all major browsers. Later allow make and craft markup grid easier

We switch the global box-sizing value to border-box. This ensures padding does not affect the final computed width of an element

#1 Update box-model to border-box 
Magento have define .lib-set-default-border-box but it's never used in code base
When use box-sizing border box we make frontend dev life easier. Only define in one file and styles apply to all other place for all element.
we can use content-box or padding-box (where supported) as well, without worrying about a universal selector overriding css

box-sizing border box defined in file app/design/frontend/Magento/blank/web/css/source/_reset.less

<img width="309" alt="Screen Shot 2020-12-03 at 10 38 15" src="https://user-images.githubusercontent.com/1908873/100960860-b9c22c00-3553-11eb-9d9d-352812cca2ae.png">

https://thisthat.dev/border-box-vs-content-box
Now a lot of web use border-box model as default

When border-box is used border,  padding-left and padding-right are not included in the calculation box width

all elements can re-define box model (content-box) without any issue if need. Default all elements in html will have border box model

This PR only remove define border-box in theme core/libs magento in luma theme
Keep old declare mixins related boxder-box for maintain compatible in 3rd-party themes
Keep backward compatible in blank theme (avoid too much changes in blank)

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Check all pages blank/luma theme make sure all page work without break, glitch, overlap box


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
Old article related
https://css-tricks.com/inheriting-box-sizing-probably-slightly-better-best-practice
Need workaround check visual both luma/blank theme when apply border-box model
Should test all screen make sure everything work like before no big visual change break any elements

CC: @krzksz @Leland @ptylek  good to hear feedback from you guys
@DanielRuf ,@Parakoopa, @kirmorozov ,@leoquijano, @jonathanKingston can you guys give me some opinion

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#31137: Update box model to border box

sidolov commented:
@magento create issue

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento give me 2.4-develop instance

mrtuvn commented:
@magento give me test instance

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

mrtuvn commented:
@krzksz @ihor-sviziev can we have priority review and approve this soon ?

mrtuvn commented:
@magento give me test instance

ihor-sviziev commented:
@mrtuvn this looks like a big change that might cause some issues during the update. I believe it should be delivered in 2.5

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml:
```diff
@@ -7,7 +7,7 @@
 <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
     <head>
-        <meta name="viewport" content="width=device-width, initial-scale=1"/>
+        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1.0, user-scalable=no"/>
```
could you explain why this change is needed?

mrtuvn replied:
As I mentioned in description 
This update meta viewport content to ensure proper rendering and touch zooming for all devices

mrtuvn replied:
i think should revert this back to origin as this change is unneccessary. Depend on projects  require we need to modify viewport in custom theme

ihor-sviziev commented on app/design/frontend/Magento/blank/Magento_Catalog/web/css/source/module/_listings.less:
```diff
@@ -34,7 +34,7 @@
 
         &-item {
-            font-size: 1.4rem;
+            .lib-font-size(14);
```
Could you explain why it's needed?

mrtuvn replied:
This is standard font mixins default of magento. It's will work same as default

ihor-sviziev commented on app/design/frontend/Magento/luma/Magento_Catalog/web/css/source/module/_toolbar.less:
```diff
@@ -170,7 +170,6 @@
     }
 
     .modes {
-        display: inline-block;
```
could you explain why it was removed?

mrtuvn replied:
When use float element will treated as block element so basically display define will unnecessary

mrtuvn replied:
Change here previously remove line will cause missing modes switch toolbar in desktop screen
I have updated that

ihor-sviziev commented on lib/web/css/source/lib/_sections.less:
```diff
@@ -17,7 +17,6 @@
         width: auto;
 
         > .switch {
-            .lib-css(height, @_tab-control-height__base);
```
could you explain why it was removed ?

mrtuvn replied:
No need fixed height here

mrtuvn replied:
![Screenshot from 2020-12-11 22-26-18](https://user-images.githubusercontent.com/1908873/101921840-036be000-3c00-11eb-8be6-00a8994a5830.png)

mrtuvn replied:
And this tab in theme blank
![Screenshot from 2020-12-13 11-27-02](https://user-images.githubusercontent.com/1908873/102003163-4f20a580-3d36-11eb-948e-39fcfffd9220.png)

ihor-sviziev commented on lib/web/css/source/lib/_sections.less:
```diff
@@ -106,7 +104,6 @@
             .lib-css(background, @_tab-control-background-color);
             .lib-css(border, @_tab-content-border);
             border-bottom: none;
-            .lib-css(height, @_tab-control-height);
```
and this one?

mrtuvn replied:
same as above

ihor-sviziev commented on lib/web/css/source/lib/_sections.less:
```diff
@@ -244,7 +239,6 @@
             .lib-css(border-left,@_accordion-control-border-left);
             .lib-css(border-right,@_accordion-control-border-right);
             .lib-css(border-top,@_accordion-control-border-top);
-            .lib-css(height, @_accordion-control-height);
```
and this

mrtuvn replied:
same as above

mrtuvn commented:
Yes i think we should improve that. Seem a lot of things will change. Can it will be a part of codebase in early realease

ihor-sviziev commented:
@mrtuvn I'll check it. Just I'm curious that such a change might break the websites during the patch Magento update.

mrtuvn commented:
From my check not much change compare with previous version(use default box-model content-box). This PR changes will apply for frontend only. Affected for both luma and blank

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
@mrtuvn ok. Let's try to deliver it in 2.4 release line. Removed 2.5 milestone

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests EE

mrtuvn commented:
Seem fails functional tests come from another modules

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

mrtuvn commented:
@magento give me test instance
CC: @ihor-sviziev

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests EE

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8552](https://jira.corp.magento.com/browse/ENGCOM-8552) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

mrtuvn commented:
Can we have QA verify for this PR

ihor-sviziev commented:
@engcom-Alfa is there any progrss?

mrtuvn commented:
rebased code
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests B2B

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests
In the mean times no tests available so i rebase + add some clean up code

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests EE

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests

ihor-sviziev commented:
@magento run Functional Tests B2B, Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Integration Tests

mrtuvn commented:
@magento give me 2.4-develop instance

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE

mrtuvn commented:
@magento run Functional Tests CE

mrtuvn commented:
@magento run all tests

Leland commented on app/design/frontend/Magento/blank/web/css/source/_reset.less:
```diff
@@ -4,5 +4,7 @@
 //  */
 
 & when (@media-common = true) {
+    .lib-html5doctor-reset();
```
This feels to me like we're including a lot of properties that we probably don't actually need. What aspects of this reset are not already being met by the generic `lib-magento-reset` mixin below it?

This line for instance: 

https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/lib/web/css/source/lib/_resets.less#L439-L442

Is mostly a dupe of this block, from `lib-magento-reset`:

https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/lib/web/css/source/lib/_resets.less#L20-L32

Perhaps rather than calling this mixin directly, add the needed lines from the html5doctor mixin to the generic reset mixin instead?

mrtuvn replied:
Thank you for your comment!! @Leland 
As frontend view can you give me more review on this PR

mrtuvn committed:
Update box model to border box
update
update
update

update styles
update
update#3


update#4


update


update box sizing listing product luma


fix static


update


update


update


up


update


update


update


update


update


update


add quotes


update


update


up


update


update


update

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests CE, Functional Tests EE, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

mrtuvn commented:
@magento run WebAPI Tests, Functional Tests B2B

ihor-sviziev commented:
@mrtuvn,
I still believe this is a too high risk of breaking frontend, so you don't expect having it in a small release, and we should post stone it to 2.5.

mrtuvn commented:
yes totally agreed. I don't think this one can be accept by internal even in 2.5. Since this one can possible introduce unexpected broken layout
